module/zfs/arc.c

```diff
@@ -1732,6 +1812,255 @@ arc_buf_try_copy_decompressed_data(arc_buf_t *buf)
+	 * already checked in arc_buf_fill() as a quick check, but we do it
+	 * again now under the hash_lock to make sure nothing has changed.
+	 * If this isn't true there is no work to do so we can simply return.
+	 */
+	if (!HDR_HAS_RABD(hdr) || hdr->b_l1hdr.b_pabd != NULL)
+		goto out_unlock;
+
+	arc_hdr_alloc_abd(hdr, B_FALSE);
+
+	/*
+	 * We must be careful to use the passed-in dsobj value here and
+	 * not the value in b_dsobj. b_dsobj is meant to be a best guess for
+	 * the L2ARC, which has the luxury of being able to fail without real
+	 * consequences (the data simply won't make it to the L2ARC). In
+	 * reality, the dsobj stored in the header may belong to a dataset
+	 * that has been unmounted or otherwise disowned, meaning the key
+	 * won't be accessible via that dsobj anymore.
+	 */
+	ret = spa_keystore_lookup_key(spa, dsobj, FTAG, &dck);
+	if (ret != 0) {
+		ret = SET_ERROR(EACCES);
+		goto error;
+	}
+
+	ret = zio_do_crypt_abd(B_FALSE, &dck->dck_key,
+	    hdr->b_crypt_hdr.b_salt, hdr->b_crypt_hdr.b_ot,
+	    hdr->b_crypt_hdr.b_iv, hdr->b_crypt_hdr.b_mac,
+	    HDR_GET_PSIZE(hdr), bswap, hdr->b_l1hdr.b_pabd,
+	    hdr->b_crypt_hdr.b_rabd, &no_crypt);
+	if (ret != 0)
+		goto error;
+
+	if (no_crypt) {
+		abd_copy(hdr->b_l1hdr.b_pabd, hdr->b_crypt_hdr.b_rabd,
+		    HDR_GET_PSIZE(hdr));
+	}
+
+	/*
+	 * If this header has disabled arc compression but the b_pabd is
+	 * compressed after decrypting it, we need to decompress the newly
+	 * decrypted data.
+	 */
+	if (HDR_GET_COMPRESS(hdr) != ZIO_COMPRESS_OFF &&
+	    !HDR_COMPRESSION_ENABLED(hdr)) {
+		/*
+		 * We want to make sure that we are correctly honoring the
+		 * zfs_abd_scatter_enabled setting, so we allocate an abd here
+		 * and then loan a buffer from it, rather than allocating a
+		 * linear buffer and wrapping it in an abd later.
+		 */
+		cabd = arc_get_data_abd(hdr, arc_hdr_size(hdr), hdr);
+		tmp = abd_borrow_buf(cabd, arc_hdr_size(hdr));
+
+		ret = zio_decompress_data(HDR_GET_COMPRESS(hdr),
+		    hdr->b_l1hdr.b_pabd, tmp, HDR_GET_PSIZE(hdr),
+		    HDR_GET_LSIZE(hdr));
+		if (ret != 0) {
+			abd_return_buf(cabd, tmp, arc_hdr_size(hdr));
+			goto error;
+		}
+
+		abd_return_buf_copy(cabd, tmp, arc_hdr_size(hdr));
+		arc_free_data_abd(hdr, hdr->b_l1hdr.b_pabd,
+		    arc_hdr_size(hdr), hdr);
+		hdr->b_l1hdr.b_pabd = cabd;
+	}
+
+	spa_keystore_dsl_key_rele(spa, dck, FTAG);
+
+out_unlock:
+	if (hash_lock != NULL)
+		mutex_exit(hash_lock);
+
+	return (0);
+
+error:
+	arc_hdr_free_abd(hdr, B_FALSE);
+	if (dck != NULL)
+		spa_keystore_dsl_key_rele(spa, dck, FTAG);
+	if (cabd != NULL)
+		arc_free_data_buf(hdr, cabd, arc_hdr_size(hdr), hdr);
+	if (hash_lock != NULL)
+		mutex_exit(hash_lock);
+
+	return (ret);
+}
+
+/*
+ * This function is used by the dbuf code to decrypt bonus buffers in place.
+ * The dbuf code itself doesn't have any locking for decrypting a shared dnode
+ * block, so we use the hash lock here to protect against concurrent writes.
```

Comment:
I mean't "memory writes", but yes that wording is probably not great. Concurrent calls to `arc_buf_fill()` qould probably be a better way to say that.